File: Practice/Dynamic_Programming/q32/q32.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int dp[500][502];
int n;

int main(void){
    cin >> n;
    for(int i = 0; i < n; i++){
        for(int j = 1; j <= i+1; j++){
            cin >> dp[i][j];
        }
    }
    for(int row = n-2; row >= 0; row--){
        for(int col = 1; col <= n; col++){
            dp[row][col] = max(dp[row][col]+dp[row+1][col], dp[row][col]+dp[row+1][col+1]);
        }
    }
    cout << dp[0][1] << "\n";
}

File: Practice/Dynamic_Programming/q32/q32.py
n = int(input())
dp = []
for _ in range(n):
    dp.append(list(map(int, input().split())))
for row in range(n-2, -1, -1):
    for col in range(row+1):
        dp[row][col] = max(dp[row][col]+dp[row+1][col], dp[row][col]+dp[row+1][col+1])
print(dp[0][0])